sycl/doc/extensions/ComplexAlgorithms/ComplexAlgorithms.asciidoc

```diff
@@ -0,0 +1,133 @@
+= SYCL_EXT_ONEAPI_COMPLEX_ALGORITHMS
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This extension adds limited support for `std::complex<float>` and
+`std::complex<double>` to several SYCL group functions and algorithms.
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Revision: 1
+
+== Contacts
+
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com)
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 4 and
+the following extensions:
+
+- https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst[C and C++ Standard libraries support]
+- https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/C-CXX-StandardLibrary/DeviceLibExtensions.rst[Device library extensions]
```

Comment:
This is probably a can of worms, but it's not clear to me that "DeviceLibExtensions.rst" really is a SYCL extension.  Do we need to list it here?

Reply:
I'm not sure either.

That document says: "Device compiler that indicates support for a particular extension is supposed to support all the corresponding functions."  My interpretation of that is that if a device compiler _doesn't_ indicate support for `cl_intel_devicelib_complex` and `cl_intel_devicelib_complex` then `std::complex<float>` and `std::complex<double>` won't work.

I do think we need to convey somehow that just because `SYCL_EXT_ONEAPI_COMPLEX_ALGORITHMS` is defined doesn't mean that all functionalities of `std::complex<>` will work -- it just means that there will be an overload of the group algorithms available for `std::complex<>`.  Would you prefer if we removed the reference to the unclear extensions, and added a paragraph stating this instead?

Reply:
OK, I just looked through the DPC++ runtime implementation, and I think I understand this better now.  The attributes like `cl_intel_devicelib_complex` are features that the device driver may or may not support.  However, the DPC++ runtime has fallback implementation for all of these features, which it uses if the device does not support it natively.  It appears that DPC++ can also "tag" a SPIR-V module with these attributes to see which features are really needed by the SPIR-V.

When the DPC++ runtime compiles a SPIR-V module for a device *D*, it checks to see which attributes the module uses and compares this with the set of attributes that the device *D* supports.  If the module uses a feature that the device does not support, it links in the fall-back implementation for that feature.

Therefore, "DeviceLibExtensions.rst" is not a SYCL extension specification, but rather it is an internal design document.  It provides a formal definition of exactly which APIs correspond to each of the `cl_intel_devicelib_*` attributes.

Moreover, the APIs listed in that document are *always* supported on every device.  Some may be supported via a fall-back implementation, but they are always supported.

Therefore, I think that a SYCL application should be able to assume that all of the `std::complex()` APIs work on any device, so long as the compiler pre-defines `SYCL_EXT_ONEAPI_COMPLEX_ALGORITHMS`.

Do you expect this extension to use only those "complex" functions that are documented in "DeviceLibExtensions.rst"?  If so, then it seems like the implementation is more straightforward.  You can just call these functions, and they are guaranteed to be implemented on the device.  However, if you plan to call any additional "complex" functions, then you will need to work with the implementation team to create fall-back implementations for them.

Reply:
> Therefore, "DeviceLibExtensions.rst" is not a SYCL extension specification, but rather it is an internal design document. It provides a formal definition of exactly which APIs correspond to each of the `cl_intel_devicelib_*` attributes.
>
> Moreover, the APIs listed in that document are _always_ supported on every device. Some may be supported via a fall-back implementation, but they are always supported.

Thanks for looking into this.  I had no idea we had a fallback implementation, and I agree it simplifies things significantly.  Do you think "DeviceLibExtensions.rst" should be lifted out of the "extensions/" directory?  Mixing extension documentation with design documents is a little confusing.

> Do you expect this extension to use only those "complex" functions that are documented in "DeviceLibExtensions.rst"? If so, then it seems like the implementation is more straightforward. You can just call these functions, and they are guaranteed to be implemented on the device. However, if you plan to call any additional "complex" functions, then you will need to work with the implementation team to create fall-back implementations for them.

My expectation is that we'll only use those functions.  All we should need, really, is the ability to construct a `std::complex` (to use as an accumulator) and the ability to perform addition.  When I spoke to the implementation team about this they assured me that all functionality from `std::complex` works, so I think we're good.

I've removed the dependency on the internal design documents in 4a4c195b80c86a70def091e09bf696fdcea62ba8.

Reply:
> Thanks for looking into this. I had no idea we had a fallback implementation, and I agree it simplifies things significantly. Do you think "DeviceLibExtensions.rst" should be lifted out of the "extensions/" directory?

Yes, definitely.  I'll address this when I reorganize the directory structure of the extension specifications.